solutions/POI 05-sza.mdx

```diff
@@ -0,0 +1,111 @@
+---
+id: poi-05-sza
+source: POI 2005
+title: Template
+author: Andi Qu
+---
+
+Let the string be $S$.
+
+<Spoiler title="Hint 1">
+
+A template of $S$ is both a prefix and a suffix of $S$. This suggests that we should use KMP.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+If $T$ is a template of $S$ and $U$ is a template of $T$, then $U$ is also a template of $S$.
+
+<Spoiler title="Proof">
+
+Since $U$ is a template of $T$, we can make $T$ using $U$. Since $T$ is a template of $S$, we can make $S$ using $T$.
+
+Therefore, we can also make $S$ using $U$ (i.e. it's a template of $S$).
+
+</Spoiler>
+
+</Spoiler>
+
+<Spoiler title="Hint 3">
+
+First, use KMP to compute the prefix function $\pi$ of $S$. The shortest template of $S$ is either $S$ itself or the shortest template of $S[0 : \pi[|S| - 1]]$.
+
+<Spoiler title="Proof">
```

Comment:
This proof seems unnecessarily long. All you need to do is show that any template of S aside from S itself is also a template of S[0 : \pi[|S| - 1]] (which you do below already, but there's no need to assume a contradiction).